### PR TITLE
Enable leap to work as a telescope item selector that triggers a pick

### PR DESCRIPTION
**Pull request by aaronlifton**
I have the following telescope mapping, which enables me to pick a single row from a Telescope list via leap (`s`). Because the mapping picks a row in telescope, it triggers the closing of the telescope window which causes leap to not be able to find the win/buf. The error can be reproduced as follows, given the following mapping:

```
local function get_telescope_targets(prompt_bufnr)
  local pick = require("telescope.actions.state").get_current_picker(prompt_bufnr)
  local scroller = require("telescope.pickers.scroller")

  local wininfo = vim.fn.getwininfo(pick.results_win)

  local first =
    math.max(scroller.top(pick.sorting_strategy, pick.max_results, pick.manager:num_results()), wininfo[1].topline - 1)
  local last = wininfo[1].botline - 1

  local targets = {}
  for row = last, first, -1 do
    local target = {
      wininfo = wininfo[1],
      pos = { row + 1, 1 },
      row = row,
      pick = pick,
    }
    table.insert(targets, target)
  end
  return targets
end

-- under telescope mappings
{
  ["s"] = function(prompt_bufnr)
    local win = vim.api.nvim_get_current_win()

    require("leap").leap({
      targets = get_telescope_targets(prompt_bufnr),
      action = function(target)
        target.pick:set_selection(target.row)
        require("telescope.actions").select_default(prompt_bufnr)
      end,
    })
  end,
}
```

So in a telescope window, entering normal mode and pressing `s` will turn the list of telescope items into leap targets, and selecting one will pick that item.
However, since this closes the telescope window, there are 2 undefined win/buf errors inside of leap, since leap seems to not expect a leap target like a telescope item. So, I protected these statements with guard clauses, and now I'm able to leap to a telescope target without error. I felt this was appropriate because it enables further usages of leap.

Here is a video of this customization in my branch working without error:

https://github.com/ggandor/leap.nvim/assets/21133757/0d8f7f90-51e7-4633-9025-732ddc271825

### CODE REVIEW

**Review by ggandor — CHANGES_REQUESTED** [review]

**Comment by ggandor** [comment]
@aaronlifton :)

File: fnl/leap/highlight.fnl [hunk 1]
```diff
@@ -21,14 +21,17 @@
 (fn M.cleanup [self affected-windows]
   ; Clear beacons & cursor.
   (each [_ [bufnr id] (ipairs self.extmarks)]
-    (api.nvim_buf_del_extmark bufnr self.ns id))
+    (let [is-valid-buf (api.nvim_buf_is_valid bufnr)]
```
[the remark this answers is not in the document]
Comment (aaronlifton) [reply fnl/leap/highlight.fnl:24]: Thanks, this was the first time I used fennel, so I assumed there were some style mistakes! Working on these changes now

File: fnl/leap/highlight.fnl [hunk 2]
```diff
@@ -21,14 +21,17 @@
...
+        (if (is-valid-buf)
+            (api.nvim_buf_del_extmark bufnr self.ns id))))
   (set self.extmarks [])
   ; Clear backdrop.
   (when (pcall api.nvim_get_hl_by_name self.group.backdrop false)  ; group exists?
     (each [_ winid (ipairs affected-windows)]
       (local wininfo (. (vim.fn.getwininfo winid) 1))
-      (api.nvim_buf_clear_namespace
-        wininfo.bufnr self.ns (dec wininfo.topline) wininfo.botline))
+      (if (> (length wininfo) 0)
```
[the remark this answers is not in the document]
Comment (aaronlifton) [reply fnl/leap/highlight.fnl]: Good call, made the changes. That's a great point, perhaps there would be `affected-bufs` too.

**Comment by ggandor** [comment]
Sorry for the delay, just did some tweaks & added you as co-author, to speed things up. Thanks very much for the contribution!

The snippet above works fine, but there is a stack overflow error if triggering it before entering any input. Do you happen to have an idea how that could be fixed? (It might worth adding it to the "Extending Leap" section.)

**Pull request closed without merging**